components/operator-service/src/main/java/com/wso2telco/dep/operatorservice/service/BlackListWhiteListService.java

```diff
@@ -291,4 +288,70 @@ public int getAPIId(String providerName, String apiName, String apiVersion) thro
 	public String[] getAPIInfo(int apiId) throws BusinessException {
 		return dao.getAPIInfo(apiId);
 	}
+
+	/**
+	 * Returns the offset and limit required for pagination of blacklisted APIs
+	 *
+	 * @param apiId apiId API ID of blacklisted numbers to get count of
+	 * @param limit maximum number of records per page
+	 * @return
+	 */
+	private List<PaginationDTO> getBlacklistPages(BlacklistWhitelistDTO dto, int limit) throws Exception {
+		int count = dao.getBlacklistCountByApi(dto);
+		double factor = (double)count / (double)limit;
+		int index = 0;
+
+		List<PaginationDTO> blacklistPages = new ArrayList<PaginationDTO>();
+
+		while (index < (int) factor) {
+			blacklistPages.add(new PaginationDTO((int) limit, (int) (index * limit)));
+			index++;
+		}
+
+		//handles records less than limit i.e: not divisible by limit
+		if (factor - (int) factor > 0.0) {
+			blacklistPages.add(new PaginationDTO((int) limit, (int) factor * (int) limit));
+		}
+
+		return blacklistPages;
+	}
+
+	public void getBlacklistAsZip(BlacklistWhitelistDTO dto, OutputStream out) throws Exception {
+		List<PaginationDTO> pages = getBlacklistPages(dto,1000000);
+		int index = 1;
+
+		ZipOutputStream zipOut = new ZipOutputStream(new BufferedOutputStream(out));
+		//for each page create new entry and write to stream
+
+		for(PaginationDTO p : pages){
+			zipOut.putNextEntry(new ZipEntry("blacklist-whitelist-numbers-" + index + ".csv"));
+			dao.streamBlacklistByApi(dto,p.getOffset(),p.getLimit(),zipOut);
+			zipOut.closeEntry();
+			index++;
+		}
+
+		zipOut.close();
+		out.flush();
+		out.close();
+	}
+
+	public BlacklistWhitelistSearchResponseDTO search(BlacklistWhitelistDTO dto) throws Exception {
+		BlacklistWhitelistSearchResponseDTO responseDTO = new BlacklistWhitelistSearchResponseDTO();
+		//validate
+		Matcher matcher = bwRegex.matcher(dto.getMsisdnList().get(0));
```

Comment:
check the msisdn list is not empty or null before accessing the 0th element and will move the validation into a common validator service

Reply:
msisdn format validation is a business logic, that's why it's been done at service layer